### PR TITLE
Add `prefer-string-raw` rule

### LINKED ISSUE #2048: Rule proposal: `prefer-string-raw`
### Description

To avoid escape `\`.

### Fail

```js
console.log('"\\" found.');
```

```js
const directory = "C:\\windows\\style\\path";
```

```js
new RegExp('foo\\.bar')
```

### Pass

```js
console.log(String.raw`"\" found.`);
```

```js
const directory = String.raw`C:\windows\style\path`;
```

```js
new RegExp(String.raw`foo\.bar`)
```

```js
const foo = '\\ `';
// Use `String.raw` will need escape `
const foo = '\\ ${foo}';
// Use `String.raw` will need escape $ or `{`
```

```js
// Places can't use a tagged template

"A directive\\"

import foo from "./foo\\bar.js";

const foo = {
	'\\': 'a non-computed key'
}
```

---

This rule can also check when building a tagged function like [this](https://github.com/fregante/code-tag/commit/b20054ce644edb85f17a93ec364371bc3943455f), but it's hard to detect such pattern.

### PR DESCRIPTION
**Pull request by fisker**
Fixes #2048

### CODE REVIEW

**Comment by fisker** [comment]
Blocked by #2342

**Comment by fisker** [comment]
Once this gets reviewed, we can revert https://github.com/sindresorhus/eslint-plugin-unicorn/pull/2339/commits/9bc7b933cbb1a34be79f4163cca1e75bb271b68e before merge.

File: docs/rules/prefer-string-raw.md
Comment (sindresorhus) [inline docs/rules/prefer-string-raw.md:10]: Should link to String.raw on MDN here.

File: docs/rules/prefer-string-raw.md
Comment (sindresorhus) [inline docs/rules/prefer-string-raw.md:1]:
```suggestion
# Prefer using the `String.raw` tag to avoid escaping `\`
```

File: docs/rules/prefer-string-raw.md
Comment (sindresorhus) [inline docs/rules/prefer-string-raw.md:22]: ^

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: rules/prefer-string-raw.js
Comment (sindresorhus) [inline rules/prefer-string-raw.js:10]: Shouldn't these use `String.raw`?

Comment (fisker) [reply rules/prefer-string-raw.js:10]: ``String.raw`\` `` is an error and ``String.raw`\\` `` products two backslash, so it's impossible to create a a single backslash with `String.raw`.

We can't do ``String.raw`foo\` `` either. That's why we ignore string ends with backslash  https://github.com/sindresorhus/eslint-plugin-unicorn/pull/2339/files#diff-def49face997bfe9729f748b6720c2847611a4ad3b6a9714a96f0ef408259d75R65

Comment (fisker) [reply rules/prefer-string-raw.js:10]: The one in `messages` can't use `String.raw` either, because it contains `` ` ``, `` String.raw`foo`bar` `` this is invalid code. Ignored in code too https://github.com/sindresorhus/eslint-plugin-unicorn/pull/2339/files#diff-def49face997bfe9729f748b6720c2847611a4ad3b6a9714a96f0ef408259d75R57

Comment (sindresorhus) [reply rules/prefer-string-raw.js:10]: Ah ok.

**Comment by sindresorhus** [comment]
This looks good to me 👍

CI needs to be fixed though.

**Pull request merged**
